src/dma_mapping.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+
+use crate::VfioContainer;
+use crate::{get_host_address_range, ExternalDmaMapping};
+use std::convert::TryInto;
+use std::io;
+use std::sync::Arc;
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+
+/// This structure implements the ExternalDmaMapping trait. It is meant to
+/// be used when the caller tries to provide a way to update the mappings
+/// associated with a specific VFIO container.
+pub struct VfioDmaMapping {
+    container: Arc<VfioContainer>,
+    memory: GuestMemoryAtomic<GuestMemoryMmap>,
```

Comment:
Is it ok to use generic type <M: GuestMemory > instead of GuestMemoryMmap here?

Reply:
Seems we should use "AS: GuestAddressSpace" to replace "GuestMemoryAtomic<GuestMemoryMmap>" here.

Reply:
We found use generic type memory which require get_host_address_range/get_region_host_address_range  all switch to the generit type, it makes some operations more complicate to convert.

Reply:
That should be the way, otherwise it may not be reused by other projects:)

Reply:
> GuestAddressSpace
we can use memory: Arc<GuestMemory>, did you mean we use memory: Arc<GuestAddressSpace> or just memory: GuestAddressSpace?